### PR TITLE
[osh] Fix regression of `${undef@P}` and `${undef@Q}` with `nounset`

### PR DESCRIPTION
**Pull request by akinomyoga**
This PR refactors and simplifies the existing code for checking `nounset`. This also fixes the regression of `${undef@P}` and `${undef@Q}` mentioned in #2186:

> https://github.com/oils-for-unix/oils/pull/2186#issuecomment-2537311633 @akinomyoga 
>
> > then change the test to
> > ```
> >            if not vsub_state.has_test_op and not vsub_state.has_nullary_op:
> >                 val = self._EmptyStrOrError(val, part.token)
> > ```
> >   
> > and then for `@Q` and `@P`, handle `case(value_e.Undef)` in the normal way, producing empty string
> 
> This change breaks `osh -uc "${undef@P}"` and `osh -uc "${undef@Q}"`. The `nounset` check is skipped.
>
> https://github.com/oils-for-unix/oils/pull/2186#issuecomment-2537349028 @andychu 
>
> Ahh we are missing test coverage for `-u` ? OK thank you, I will look at it
>
> https://github.com/oils-for-unix/oils/pull/2186#issuecomment-2537384972 @akinomyoga 
>
> As I mentioned, in my local repository, I already have commits to fix everything by refactoring the calls of `_EvalBracketOp`.
>
> https://github.com/oils-for-unix/oils/pull/2186#issuecomment-2537419550 @andychu 
>
> Ah OK, yes definitely feel free to fix it along with the other issues!

### CODE REVIEW

File: osh/word_eval.py [hunk 1]
```diff
@@ -1461,12 +1465,18 @@ def _EvalBracedVarSub(self, part, part_vals, quoted):
...
-                              op_tok)
+                    # undef -> empty array
+                    val = self._ProcessUndef(val, part.token, vsub_state)
+
+                    # ${!a[@]-'default'} is a non-fatal runtime error in bash.
+                    # Here it's fatal.
+                    if suffix_op_ and suffix_op_.tag() == suffix_op_e.Unary:
+                        unary_op = cast(suffix_op.Unary, suffix_op_)
+                        if consts.GetKind(unary_op.op.id) == Kind.VTest:
+                            op_tok = cast(suffix_op.Unary, UP_op).op
```
Comment (andychu) [inline osh/word_eval.py:1476]: Are op_tok and unary_op the same variable?    Can we simplify it and just use `unary_op` , and get rid of op_tok ?


I think there was some existing confusion around the suffix_op_ var and UP_op already in the code

the `UP_op` variable is intended to be used for downcasting, e.g. after `wth tagswitch()`, although it's not a hard rule

**Review by andychu — COMMENTED** [review]
Thanks for this fix!   And the code looks cleaner in general

File: osh/word_eval.py (hunk 1 above)
Comment (akinomyoga) [reply osh/word_eval.py]: Yeah, good point. I fixed it in 39ca6a61aef1c4a9e76ccc41b679d46ed7748a83..fc98f1a5f4243f4c5139f20b2288760680208437

File: osh/word_eval.py [hunk 2]
```diff
@@ -1266,25 +1279,26 @@ def _DecayArray(self, val):
         tmp = [s for s in bash_impl.BashArray_GetValues(val) if s is not None]
         return value.Str(sep.join(tmp))
 
-    def _EmptyStrOrError(self, val, token):
-        # type: (value_t, Token) -> value_t
+    def _ProcessUndef(self, val, token, vsub_state):
```
[the remark this answers is not in the document]
Comment (akinomyoga) [reply osh/word_eval.py]: I applied also this change in 39ca6a61aef1c4a9e76ccc41b679d46ed7748a83..fc98f1a5f4243f4c5139f20b2288760680208437.

File: osh/word_eval.py (hunk 1 above)
Comment (akinomyoga) [reply osh/word_eval.py]: Additional fix: fc98f1a5f4243f4c5139f20b2288760680208437..ce4dc6462a3f5572021e46d2b7168202a4322721

**Comment by andychu** [comment]
Great thanks!

I temporarily broke the soil-staging branch, I will fix it

**Pull request merged**